CUDA support in the CSR layout: constructors

aocsa opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56561 CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr
* **#56560 CUDA support in the CSR layout: constructors**
* #56744 Fix: Avoid unnecessary creation of tensors in SparseCsrTensorImpl::resize_and_clear_
* #56723 Fix: sparse_csr_tensor segfaults when crow_indices or col_indices are non-tensors
* #56722 Fix: Incorrect example output in sparse_csr_tensor doc-string

Closes gh-56485

Differential Revision: [D27939681](https://our.internmc.facebook.com/intern/diff/D27939681)

aocsa committed:
CUDA support in the CSR layout: constructors

[ghstack-poisoned]

aocsa committed:
Update on "CUDA support in the CSR layout: constructors"




[ghstack-poisoned]

aocsa committed:
Update on "CUDA support in the CSR layout: constructors"


Closes gh-56485



[ghstack-poisoned]

aocsa committed:
Update on "CUDA support in the CSR layout: constructors"


Closes gh-56485



[ghstack-poisoned]

aocsa committed:
Update on "CUDA support in the CSR layout: constructors"


Closes gh-56485



[ghstack-poisoned]

ezyang commented on torch/testing/_internal/common_utils.py:
```diff
@@ -1049,6 +1049,46 @@ def setUp(self):
+                    r[last_i + 1] += 1
+            for _i in range(last_i, dim):
+                r[_i + 1] = r[last_i + 1]
+            return r
```
For your testing sanity you probably are going to want to sub in a more efficient implementation of this so that your sparse tests don't take foreeever to run

ezyang commented on aten/src/ATen/native/sparse/SparseCsrTensor.cpp:
```diff
@@ -25,7 +25,6 @@ SparseCsrTensor new_csr_tensor(const TensorOptions& options) {
   if (options.device().is_cuda()) {
     dispatch_key = DispatchKey::SparseCsrCUDA;
   } else {
-    TORCH_INTERNAL_ASSERT(options.device().is_cpu());
```
This assert still seems valid...

aocsa replied:
I removed this assert because is too weak and added this instead https://github.com/pytorch/pytorch/blob/dd45bcfba054cedefbcc558c2b6a4859d322c3ff/aten/src/ATen/native/sparse/SparseCsrTensor.cpp#L49   

It covers cases when options.device() is kMeta for example.

ezyang commented:
I kind of don't want to take this PR while there is a flagrantly slow CSR sample input implementation, this is the sort of thing that never gets fixed in the future. What do you think?

aocsa committed:
Update on "CUDA support in the CSR layout: constructors"


Closes gh-56485

Differential Revision: [D27939681](https://our.internmc.facebook.com/intern/diff/D27939681)

[ghstack-poisoned]

aocsa committed:
Update on "CUDA support in the CSR layout: constructors"


Closes gh-56485

Differential Revision: [D27939681](https://our.internmc.facebook.com/intern/diff/D27939681)

[ghstack-poisoned]

aocsa commented:
> I kind of don't want to take this PR while there is a flagrantly slow CSR sample input implementation, this is the sort of thing that never gets fixed in the future. What do you think?


I updated the CSR sample input generator rewriting the `coo_to_csr` function and using vectorized apis. 
 https://github.com/pytorch/pytorch/blob/e251147ea51c123fc52a11f9ab9f34794ea15353/torch/testing/_internal/common_utils.py#L1061

---
<details>
<summary> new impl (expand for details).
</summary> 
- `new`

```
[----------------------------- sparse.mm -----------------------------]
                          |  gen_sparse_csr CPU  |  gen_sparse_csr CUDA
1 threads: ------------------------------------------------------------
      (32, 100) | 0.1     |          4.2         |           18.6      
      (32, 100) | 0.25    |         10.3         |           46.5      
      (512, 100) | 0.1    |         65.6         |          294.8      
      (512, 100) | 0.25   |        163.9         |          737.7      
      (1024, 100) | 0.1   |        131.1         |          584.7      
      (1024, 100) | 0.25  |        329.3         |         1526.2      
      (2048, 100) | 0.1   |        276.6         |         1182.5      
      (2048, 100) | 0.25  |        655.8         |         3267.8      

Times are in milliseconds (ms).
```
</details>
---
<details><summary> old impl (expand for details).
</summary> 
- `old`

 
```
[----------------------------- sparse.mm -----------------------------]
                          |  gen_sparse_csr CPU  |  gen_sparse_csr CUDA
1 threads: ------------------------------------------------------------
      (32, 100) | 0.1     |           7.6        |           36.7      
      (32, 100) | 0.25    |          17.9        |           84.5      
      (512, 100) | 0.1    |         115.1        |          649.8      
      (512, 100) | 0.25   |         287.3        |         1383.0      
      (1024, 100) | 0.1   |         227.9        |         1147.7      
      (1024, 100) | 0.25  |         574.0        |         2820.4      
      (2048, 100) | 0.1   |         465.7        |         2301.0      
      (2048, 100) | 0.25  |        1208.4        |         5940.8      

```
</details>

This version on average is 60% faster than previous version. Let me know your thoughts cc @ezyang
Note: I used this [script](https://gist.github.com/aocsa/9594f45b336065e4f887006110f0df2b
) to check performance.

aocsa committed:
Update on "CUDA support in the CSR layout: constructors"


Closes gh-56485

Differential Revision: [D27939681](https://our.internmc.facebook.com/intern/diff/D27939681)

[ghstack-poisoned]

aocsa committed:
Update on "CUDA support in the CSR layout: constructors"


Closes gh-56485

Differential Revision: [D27939681](https://our.internmc.facebook.com/intern/diff/D27939681)

[ghstack-poisoned]

aocsa committed:
Update on "CUDA support in the CSR layout: constructors"


Closes gh-56485

Differential Revision: [D27939681](https://our.internmc.facebook.com/intern/diff/D27939681)

[ghstack-poisoned]

aocsa committed:
Update on "CUDA support in the CSR layout: constructors"


Closes gh-56485

Differential Revision: [D27939681](https://our.internmc.facebook.com/intern/diff/D27939681)

[ghstack-poisoned]

aocsa committed:
Update on "CUDA support in the CSR layout: constructors"


Closes gh-56485

Differential Revision: [D27939681](https://our.internmc.facebook.com/intern/diff/D27939681)

[ghstack-poisoned]

ezyang commented:
I guess these gonna have to be restacked on the new stack. I'll hold off reviewing this in case you end up closing and opening new PRs.

aocsa commented:
> I guess these gonna have to be restacked on the new stack. I'll hold off reviewing this in case you end up closing and opening new PRs.

Yes, I soon as I have ready the new stack I will ping you. Thanks @ezyang
